src/relax/training/utils.cc

```diff
@@ -0,0 +1,177 @@
+    // mutate only the last block.
+    Array<BindingBlock> blocks;
+    for (int i = 0; i < static_cast<int>(seq_expr->blocks.size()); ++i) {
+      CHECK(seq_expr->blocks[i].as<DataflowBlockNode>())
+          << "All blocks in original functions should be Dataflow Block.";
+      if (i < static_cast<int>(seq_expr->blocks.size()) - 1) {
+        blocks.push_back(seq_expr->blocks[i]);
+      } else {
+        BindingBlock new_block = this->VisitBindingBlock(seq_expr->blocks[i]);
+        if (!new_block->bindings.empty()) {
+          blocks.push_back(new_block);
+        }
+      }
+    }
+    return SeqExpr(blocks, loss_body_->body);
+  }
+
+  BindingBlock VisitBindingBlock_(const DataflowBlockNode* block) override {
+    builder_->BeginDataflowBlock();
+    // emit original bindings.
+    for (const auto& binding : block->bindings) {
+      this->VisitBinding(binding);
+    }
+
+    ICHECK(orig_rets_var_.size() == orig_rets.size());
+    for (int i = 0; i < static_cast<int>(orig_rets_var_.size()); ++i) {
+      if (orig_rets_var_[i].defined()) {
+        builder_->EmitNormalized(VarBinding(orig_rets_var_[i].value(), orig_rets[i]));
+      }
+    }
+
+    // emit blocks for loss function part.
+    for (const BindingBlock& block : loss_body_->blocks) {
+      CHECK(block.as<DataflowBlockNode>())
+          << "All blocks in loss functions should be Dataflow Block.";
+      for (const Binding& binding : block->bindings) {
+        this->VisitBinding(binding);
+      }
+    }
+
+    return builder_->EndBlock();
+  }
+
+  void VisitBinding_(const VarBindingNode* binding) override {
+    Var new_var = Downcast<Var>(this->VisitExpr(binding->var));
+    Expr new_value = this->VisitExpr(binding->value);
+    builder_->EmitNormalized(VarBinding(new_var, new_value));
+  }
+
+  // remap orignal dataflow var
+  // TODO(chaofan): a better way to check whether new_ret_var should be dataflow
+  void RemapToDataflow(SeqExpr body) {
+    for (const BindingBlock& block : body->blocks) {
+      for (const Binding& binding : block->bindings) {
+        const auto* binding_node = binding.as<VarBindingNode>();
+        if (binding_node && !binding_node->var->IsInstance<DataflowVarNode>()) {
+          Var new_binding_var = DataflowVar(
+              binding_node->var->vid, GetStructInfo(binding_node->var), binding_node->var->span);
+          this->var_remap_[binding_node->var->vid] = new_binding_var;
+        }
+      }
+    }
+  }
+
+  Array<Var> RemapLossParams(const Array<Var>& loss_func_params, Array<Var> new_params) {
+    for (int i = 0; i < static_cast<int>(loss_func_params.size()); ++i) {
+      Var loss_param = loss_func_params[i];
+      auto loss_param_sinfo = GetStructInfo(loss_param);
+
+      if (i < static_cast<int>(orig_rets.size())) {  // map return value to loss param
+        auto orig_ret_sinfo = GetStructInfo(orig_rets[i]);
+        ICHECK(StructuralEqual()(orig_ret_sinfo, loss_param_sinfo))
```

Comment:
No need to create an instance each time. You can have a `StructuralEqual equal` field in the class like https://github.com/mlc-ai/relax/blob/55c9fb93c8b72f62d5a677e29bdad10bcf65e737/src/relax/transform/fuse_tir.cc#L155-L156